shared/middleware/files_api.rb

```diff
@@ -51,6 +51,48 @@ def can_view_abusive_assets?(encrypted_channel_id)
     teaches_student?(owner_user_id)
   end
 
+  def report_too_large(quota_type)
+    # Don't record a custom event since these events may be very common.
+    increment_metric('FileTooLarge', quota_type)
+    too_large
+  end
+
+  def quota_half_used?(app_size, body_length)
+    app_size < FilesApi::max_app_size / 2 && app_size + body_length >= FilesApi::max_app_size / 2
+  end
+
+  def report_quota_half_used(quota_type, encrypted_channel_id)
+    quota_event_type = 'QuotaHalfUsed'
+    increment_metric(quota_event_type, quota_type)
+    record_event(quota_event_type, quota_type, encrypted_channel_id)
+  end
```

Comment:
I am not too worried about crossing multiple times, since it does seem unlikely. If this happens, we can dig into the custom events log to see if a bunch of events were generated by the same user/app.